jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,660 @@
+   * @returns absolute price rel_tick price levels above px_base
+   */
+  price4_t
+  relative_above_base(price4_t const& px_base, int const rel_tick) const {
+    int const wire_max = max_price_field_value<price4_t>().as_integer();
+
+    // check if px_base is above $1.00
+    if (px_base > price4_t(TK_DOLLAR)) {
+      // already better than $1.00, then rel_tick is in penny
+      int wire_new = px_base.as_integer() + 100 * rel_tick;
+      wire_new = (wire_new < wire_max) ? wire_new : wire_max;
+      return price4_t(wire_new);
+    }
+    // px_base is below $1.00...
+    // ... get the ticks to reach $1.00
+    int tk_mill = price_levels(px_base, price4_t(TK_DOLLAR));
+    // ... verify if rel_tick is not enough to reach it
+    if (rel_tick < tk_mill) {
+      // rel_tick is in mills
+      return px_base + price4_t(rel_tick);
+    }
+    // get the ticks above $1.00
+    int tk_penny = rel_tick - tk_mill;
+    return price4_t(TK_DOLLAR + tk_penny * 100);
+  }
+
+  /**
+   * @param px_base absolute base price
+   * @param rel_tick price levels below px_base
+   * @returns an absolute price that is rel_tick price levels below px_base
+   */
+  price4_t
+  relative_below_base(price4_t const& px_base, int const rel_tick) const {
+    // check if px_base is above $1.00
+    if (px_base < price4_t(TK_DOLLAR)) {
+      // already below than $1.00, then rel_tick is in mill
+      int rel_base = px_base.as_integer();
+      rel_base = (rel_base < rel_tick) ? 0 : rel_base - rel_tick;
+      return price4_t(rel_base);
+    }
+    // px_base is above $1.00...
+    // ... get the ticks to reach $1.00
+    int tk_penny = price_levels(price4_t(TK_DOLLAR), px_base);
+    // ... verify if rel_tick is not enough to reach it
+    if (rel_tick < tk_penny) {
+      // rel_tick is in penny
+      int rel_base = px_base.as_integer() - rel_tick * 100;
+      return price4_t(rel_base);
+    }
+    // get the ticks below $1.00
+    int tk_mill = rel_tick - tk_penny;
+    int rel_base = px_base.as_integer();
+    rel_base = (tk_mill > TK_DOLLAR) ? 0 : TK_DOLLAR - tk_mill;
+    return price4_t(rel_base);
+  }
+
+  /**
+   * Transforms an absolute price into a top_levels_ relative position.
+   *
+   * @throw feed_error px has to be better or equal to px_begin_top_
+   *
+   * @param px an absolute price
+   * @returns top_levels_ relative position of px compare with px_begin_top_
+   */
+  std::size_t price_to_relative(price4_t const& px) const {
+    if (better_(px_begin_top_, px)) {
+      std::ostringstream os;
+      os << "array_based_order_book::price_to_relative."
+         << " Price out of range"
+         << " px_begin_top_=" << px_begin_top_ << " px=" << px;
+      throw jb::feed_error(os.str());
+    }
+    int tk_px = price_levels(price4_t(0), px);
+    int tk_begin_top = price_levels(price4_t(0), px_begin_top_);
+    return std::abs(tk_px - tk_begin_top);
+  }
+
+  /**
+   * @throw feed_error top_levels_ is empty
```

Comment:
Please document the normal case (return blah) before the exceptional case (throw bleh)